docs/docker_env_vars.md
# Docker environment variables

## Basic setup

### `BASE_URL` (required)

The base URL where the app is accessed (all others will be blocked).
Will be used to generate QR code URLs (syntax: `<BASE_URL>/manage/<uuid>`).
Should not include protocol.

Example: `plants.joshmedeiros.dev`

### `SECRET_KEY` (required)

A random string used for the django secret key.
If this is not set a random string will be generated each time the container starts (not recommended, will cause issues with session cookies).

### `DATABASE_PASSWORD` (required)

Postgres database password, must match the `plant-tracker-db` container's `POSTGRES_PASSWORD` env var.
There is NO DEFAULT for this variable, if it is not set (or does not match postgres) the app will not start.

### `SINGLE_USER_MODE` (optional)

Disables user accounts and authentication if True.
If not set user accounts will be enabled (requires registration + login).

### `DJANGO_SUPERUSER_USERNAME` (optional)

Username for the django superuser (admin control panel).
If not set the superuser will not be created.

### `DJANGO_SUPERUSER_PASSWORD` (optional)

Password for the django superuser (admin control panel).
If not set the superuser will not be created.

### `DJANGO_SUPERUSER_EMAIL` (optional)

Email for the django superuser (admin control panel).
The superuser can be created with no email (only requires username and password).

### `GUNICORN_WORKERS` (optional)

The number of gunicorn workers to use (defaults to 2 if not set).



## Email (optional)

Public deployments should include SES credentials to send user verification emails.
See [deployment.md](docs/deployment.md#Email) for instructions.

### `SES_SMTP_USER`

Amazon SES SMTP username.

### `SES_SMTP_PASSWORD`

Amazon SES SMTP password.

### `DEFAULT_FROM_EMAIL`

Email address users will see as the sender.

Example: `support@plants.joshmedeiros.dev`



## Static file CDN (optional)

Cloudfront will be used to serve static files (JS + CSS bundles) if the variable below is set.

### `STATIC_URL`

The full cloudfront distribution URL including protocol.

Example: `https://static.plants.joshmedeiros.dev`



## S3 storage (optional)

All variables below must be set to store user-uploaded photos in AWS S3.

### `AWS_ACCESS_KEY_ID`

AWS IAM user public access key ID.
User must have `AmazonS3FullAccess` permission policy (read/write S3 access).
AWS -> IAM -> Access management -> Users -> Create user

### `AWS_SECRET_ACCESS_KEY`

AWS IAM user private access key ID (only shown once when created).
User must have `AmazonS3FullAccess` permission policy (read/write S3 access).
AWS -> IAM -> Access management -> Users -> Create user

### `AWS_STORAGE_BUCKET_NAME`

Name of the S3 bucket to store photos.

### `AWS_S3_REGION_NAME`

Name of the AWS region where the S3 bucket is located.

Example: `us-west-2`

### `IMAGE_URL`

Cloudfront distribution URL or alternate domain name, excluding protocol.
Can be the same as `STATIC_URL` (minus protocol) or different subdomain.

Example: `media.plants.joshmedeiros.dev`

### `CLOUDFRONT_KEY_ID`

Cloudfront key group ID, used for signed cookies (photo access control).
See [deployment.md](docs/deployment.md) for instructions.
Must generate rsa key and upload public key to AWS. The `private_key.pem` must be mounted into the docker container at `/mnt/backend/private_key.pem`.
- If a different location is used the `CLOUDFRONT_PRIVKEY_PATH` env var must be set to that location.



## Database + cache overrides (development only)

The variables below can be used to override the default postgres and redis configuration.

These usually DO NOT need to be set, the default [docker-compose.yaml](docker/docker-compose.yaml) comes with postgres and redis containers configured for the defaults.

These are mostly only useful to connect to non-docker postgres and redis instances (development).

### `DATABASE_HOST`

Postgres database host.
Docker image defaults to `plant-tracker-db` if not set (postgres container name).
When running outside docker this defaults to `127.0.0.1` (local postgres instance).

### `DATABASE_PASSWORD`

Postgres database password - no default, app will not work without setting this.

### `DATABASE_NAME`

Postgres database name (defaults to `plant_tracker` if not set).

### `DATABASE_USER`

Postgres database username (defaults to `postgres` if not set).

### `DATABASE_PORT`

Postgres database port (defaults to `5432` if not set).

### `REDIS_HOST`

Redis host.
Docker image defaults to `plant-tracker-redis` if not set (redis container name).
When running outside docker this defaults to `127.0.0.1` (local redis instance).

### `REDIS_PORT`

Redis port (defaults to `6379` if not set).


backend/backend/settings.py
"""
Django settings for backend project.

Generated by 'django-admin startproject' using Django 5.0.1.

For more information on this file, see
https://docs.djangoproject.com/en/5.0/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/5.0/ref/settings/
"""

import os
import sys
import json
from pathlib import Path
from django.core.exceptions import ImproperlyConfigured
from django.core.management.utils import get_random_secret_key

from .validate_url_prefix import validate_url_prefix

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/5.0/howto/deployment/checklist/

# Read SECRET_KEY from env var, or generate new key if not present
SECRET_KEY = os.environ.get('SECRET_KEY')
if SECRET_KEY is None:
    SECRET_KEY = get_random_secret_key()

# Read BASE_URL from env var
BASE_URL = os.environ.get('BASE_URL')
if not BASE_URL:
    raise ImproperlyConfigured('BASE_URL is required')

# Use BASE_URL for ALLOWED_HOSTS and CSRF_TRUSTED_ORIGINS
ALLOWED_HOSTS = [BASE_URL]
CSRF_TRUSTED_ORIGINS = [f'http://{BASE_URL}', f'https://{BASE_URL}']

# Get URL prefix used to generate QR code stickers
URL_PREFIX = validate_url_prefix(BASE_URL)

# Redirect to overview page after successful login
LOGIN_REDIRECT_URL="/"

# Read SINGLE_USER_MODE from env var, or default to False if not present
# If True authentication is disabled, all plants are owned by DEFAULT_USERNAME
# If False authentication is required, separate accounts own separate plants
SINGLE_USER_MODE = os.environ.get('SINGLE_USER_MODE', '').lower() == 'true'
DEFAULT_USERNAME='DEFAULT'

# Email settings when not running in single user mode (verify account emails)
if not SINGLE_USER_MODE:
    EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
    EMAIL_HOST = 'email-smtp.us-west-2.amazonaws.com'
    EMAIL_PORT = 587
    EMAIL_USE_TLS = True
    EMAIL_HOST_USER = os.environ.get('SES_SMTP_USER')
    EMAIL_HOST_PASSWORD = os.environ.get('SES_SMTP_PASSWORD')
    DEFAULT_FROM_EMAIL = os.environ.get('DEFAULT_FROM_EMAIL', f'support@{BASE_URL}')

# Application definition

INSTALLED_APPS = [
    "plant_tracker.apps.PlantTrackerConfig",
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    # Enforce unique constraint on user email field
    "unique_user_email"
]

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
]

# Use cached sessions (fallback to database if session no longer in redis)
SESSION_ENGINE = "django.contrib.sessions.backends.cached_db"

# Password validation
# https://docs.djangoproject.com/en/5.0/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
]

# Allow logging in with username or email
AUTHENTICATION_BACKENDS = [
    "unique_user_email.backend.EmailBackend",
    "django.contrib.auth.backends.ModelBackend",
]

ROOT_URLCONF = "backend.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [ BASE_DIR / 'templates' ],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "backend.wsgi.application"


# ===============================
# Database
# ===============================
# https://docs.djangoproject.com/en/5.0/ref/settings/#databases

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": os.environ.get('DATABASE_NAME', "plant_tracker"),
        "USER": os.environ.get('DATABASE_USER', "postgres"),
        "PASSWORD": os.environ.get('DATABASE_PASSWORD'),
        "HOST": os.environ.get('DATABASE_HOST', "127.0.0.1"),
        "PORT": os.environ.get('DATABASE_PORT', "5432"),
    }
}

# Default primary key field type
# https://docs.djangoproject.com/en/5.0/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"


# ===============================
# Caching
# ===============================

REDIS_HOST = os.environ.get('REDIS_HOST', "127.0.0.1")
REDIS_PORT = os.environ.get('REDIS_PORT', "6379")
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": f"redis://{REDIS_HOST}:{REDIS_PORT}/1",
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient",
        }
    }
}

# Celery settings
CELERY_BROKER_URL = f"redis://{REDIS_HOST}:{REDIS_PORT}"
CELERY_RESULT_BACKEND = f"redis://{REDIS_HOST}:{REDIS_PORT}"
CELERY_BROKER_CONNECTION_RETRY_ON_STARTUP = True


# ===============================
# Debug mode
# ===============================

# SECURITY WARNING: don't run with debug turned on in production!
# Disable debug unless env var set
DEBUG = bool(os.environ.get('DEBUG_MODE', 0))
# User-configurable debug tool (django-debug-toolbar or silk, default to silk)
DEBUG_TOOL = os.environ.get('DEBUG_TOOL', 'silk')

# Disable non-SSL connections (except in debug mode)
if not DEBUG:
    # Only send cookies with SSL
    CSRF_COOKIE_SECURE = True
    SESSION_COOKIE_SECURE = True
    # Block all non-https requests (1 year expiration time)
    SECURE_HSTS_PRELOAD = True
    SECURE_HSTS_SECONDS = 3600
    # SECURE_HSTS_SECONDS = 31536000
    SECURE_HSTS_INCLUDE_SUBDOMAINS = True
    # Prevent browsers guessing content type (eg user uploads malicious JS as
    # photo, browser sees contents and tries to execute it)
    SECURE_CONTENT_TYPE_NOSNIFF = True

# Running tests or pylint
TESTING = 'test' in sys.argv or 'pylint' in str(sys.argv)

# Add debug tools in debug mode
if DEBUG and not TESTING:
    # Allow connections on all hosts in debug mode
    ALLOWED_HOSTS = ['*']

    # Add django-debug-toolbar if env var set
    if DEBUG_TOOL.lower() in ("toolbar", "debug_toolbar"):
        INSTALLED_APPS.append("debug_toolbar")
        MIDDLEWARE.append("debug_toolbar.middleware.DebugToolbarMiddleware")
        # Make it work on all IPs (won't work in docker dev setup otherwise)
        DEBUG_TOOLBAR_CONFIG = {
            'SHOW_TOOLBAR_CALLBACK': lambda request: True,
        }
    # Otherwise add django-silk
    else:
        INSTALLED_APPS.append("silk")
        MIDDLEWARE.append("silk.middleware.SilkyMiddleware")


# ===============================
# User-uploaded photos
# ===============================

# Serve files from local media root if any required AWS S3 env vars are missing
LOCAL_MEDIA_ROOT = not all(
    var in os.environ
    for var in [
        "IMAGE_URL",
        "AWS_ACCESS_KEY_ID",
        "AWS_SECRET_ACCESS_KEY",
        "AWS_STORAGE_BUCKET_NAME",
        "AWS_S3_REGION_NAME",
        "CLOUDFRONT_KEY_ID",
    ]
)

# Local media root settings
if LOCAL_MEDIA_ROOT:
    # Create data directory if it doesn't exist
    if not os.path.isdir(os.path.join(BASE_DIR, 'data')):
        os.mkdir(os.path.join(BASE_DIR, 'data'))

    # Create image storage directory if it doesn't exist
    if not os.path.isdir(os.path.join(BASE_DIR, 'data', 'images')):
        os.mkdir(os.path.join(BASE_DIR, 'data', 'images'))

    # Set image storage directory and URL
    MEDIA_ROOT = os.path.join(BASE_DIR, 'data', 'images')
    MEDIA_URL = "/media/"

    STORAGES = {
        # User-uploaded photos
        "default": {
            "BACKEND": "django.core.files.storage.FileSystemStorage",
            "OPTIONS": {
                "location": MEDIA_ROOT,
                "base_url": MEDIA_URL,
            },
        },
        "staticfiles": {
            "BACKEND": "whitenoise.storage.CompressedManifestStaticFilesStorage",
        },
    }

# AWS S3 settings
else:
    # Read AWS settings from env vars
    IMAGE_URL = os.environ.get("IMAGE_URL")
    AWS_ACCESS_KEY_ID = os.environ.get("AWS_ACCESS_KEY_ID")
    AWS_SECRET_ACCESS_KEY = os.environ.get("AWS_SECRET_ACCESS_KEY")
    AWS_STORAGE_BUCKET_NAME = os.environ.get("AWS_STORAGE_BUCKET_NAME")
    AWS_S3_REGION_NAME = os.environ.get("AWS_S3_REGION_NAME")
    CLOUDFRONT_KEY_ID = os.environ.get("CLOUDFRONT_KEY_ID")
    CLOUDFRONT_PRIVKEY_PATH = os.environ.get("CLOUDFRONT_PRIVKEY_PATH", "./private_key.pem")

    # Add middleware that sets cloudfront signed cookies
    # Must be after auth middleware (checks if user is authenticated)
    MIDDLEWARE.insert(
        MIDDLEWARE.index("django.contrib.auth.middleware.AuthenticationMiddleware") + 1,
        "backend.CloudFrontCookieMiddleware.CloudFrontCookieMiddleware"
    )

    STORAGES = {
        # User-uploaded photos (requires signed cloudfront cookies)
        "default": {
            "BACKEND": "storages.backends.s3.S3Storage",
            "OPTIONS": {
                "bucket_name": AWS_STORAGE_BUCKET_NAME,
                "region_name": AWS_S3_REGION_NAME,
                "custom_domain": IMAGE_URL,
                # Used signed cookies (not querystring auth)
                "querystring_auth": False,
                # Cache up to 30 days
                "object_parameters": {
                    "CacheControl": "public,max-age=2592000,immutable"
                },
            },
        },
        "staticfiles": {
            "BACKEND": "whitenoise.storage.CompressedManifestStaticFilesStorage",
        },
    }

    # Set base URL used to build photo URLs
    MEDIA_URL = f"https://{AWS_STORAGE_BUCKET_NAME}.s3.{AWS_S3_REGION_NAME}.amazonaws.com/"

# Read photo thumbnail/preview resolution and quality env vars, or use defaults
try:
    px = int(os.environ.get('THUMBNAIL_RESOLUTION', 200))
    THUMBNAIL_RESOLUTION = (px, px)
except ValueError as exc:
    raise ImproperlyConfigured('THUMBNAIL_RESOLUTION must be an integer') from exc
try:
    THUMBNAIL_QUALITY = int(os.environ.get('THUMBNAIL_QUALITY', 65))
except ValueError as exc:
    raise ImproperlyConfigured('THUMBNAIL_QUALITY must be an integer') from exc
try:
    px = int(os.environ.get('PREVIEW_RESOLUTION', 800))
    PREVIEW_RESOLUTION = (px, px)
except ValueError as exc:
    raise ImproperlyConfigured('PREVIEW_RESOLUTION must be an integer') from exc
try:
    PREVIEW_QUALITY = int(os.environ.get('PREVIEW_QUALITY', 80))
except ValueError as exc:
    raise ImproperlyConfigured('PREVIEW_QUALITY must be an integer') from exc


# ===============================
# Internationalization
# ===============================
# https://docs.djangoproject.com/en/5.0/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "UTC"

USE_I18N = True

USE_TZ = True


# ===============================
# Static files (CSS, JavaScript, Images)
# ===============================
# https://docs.djangoproject.com/en/5.0/howto/static-files/

# Serve static files from CDN if configured
STATIC_HOST = os.environ.get("STATIC_URL", "")

STATIC_URL = STATIC_HOST + "/static/"

STATIC_ROOT = BASE_DIR / "staticfiles"
